src/utils/allocators.zig

```diff
@@ -2,6 +2,251 @@ const std = @import("std");
+    collapse_sleep_ms: u32 = 100,
+    min_split_size: u64 = 128,
+}) type {
+    std.debug.assert(config.min_split_size > 0);
+
+    return struct {
+        // records are used to keep track of the memory blocks
+        records: std.ArrayList(Record),
+        // memory holds blocks of memory ([]T) that can be allocated/deallocated
+        memory: std.ArrayList([]T),
+        // allocator used to alloc the memory blocks
+        memory_allocator: std.mem.Allocator,
+        capacity: u64,
+        // for thread safety
+        mux: std.Thread.Mutex = .{},
+
+        // NOTE: we use the global_index to support fast loading the state
+        pub const Record = struct {
+            is_free: bool,
+            buf: []T,
+            global_index: u64,
+            len: u64,
+            // NOTE: this is tracked for correct usage of collapse()
+            memory_index: u64,
+
+            /// NOTE: we dont want to re-write the buf memory (that exists in `memory` field) on save
+            pub const @"!bincode-config:buf": sig.bincode.FieldConfig([]T) = .{
+                .skip = true,
+                .default_value = &.{},
+            };
+        };
+        const Self = @This();
+
+        const AllocatorConfig = struct {
+            memory_allocator: std.mem.Allocator,
+            records_allocator: std.mem.Allocator,
+        };
+
+        pub fn init(allocator_config: AllocatorConfig) Self {
+            return .{
+                .records = std.ArrayList(Record).init(allocator_config.records_allocator),
+                .memory = std.ArrayList([]T).init(allocator_config.records_allocator),
+                .memory_allocator = allocator_config.memory_allocator,
+                .capacity = 0,
+            };
+        }
+
+        pub fn create(allocator_config: AllocatorConfig) !*Self {
+            const self = try allocator_config.records_allocator.create(Self);
+            self.* = Self.init(allocator_config);
+            return self;
+        }
+
+        pub fn deinit(self: *Self) void {
+            if (config.thread_safe) self.mux.lock();
+            defer if (config.thread_safe) self.mux.unlock();
+
+            for (self.memory.items) |block| {
+                self.memory_allocator.free(block);
+            }
+            self.memory.deinit();
+            self.records.deinit();
+        }
+
+        /// append a block of N elements to the manager
+        pub fn append(self: *Self, n: u64) !void {
+            if (n == 0) return;
+
+            if (config.thread_safe) self.mux.lock();
+            defer if (config.thread_safe) self.mux.unlock();
+
+            const buf = try self.memory_allocator.alloc(T, n);
+            // NOTE: we do this here so bincode serialization can work correctly
+            // otherwise, we run into undefined memory which breaks bincode
+            @memset(buf, T.DEFAULT);
+
+            try self.records.append(.{
+                .is_free = true,
+                .buf = buf,
```

Comment:
more of a question, i believe the `buf` created on [line 86](https://github.com/Syndica/sig/pull/381/files#diff-28acc4df26271c03a683bc2a56dc0ae9aa1199fb8b2732b93aa247ae89e3eaa6R86) is being copied to the struct here, and also copied in the `try self.memory.append(buf)` call on line 98

So one alloc of `buf` at line 86 will lead to consuming 2x of the space occupied by `buf`. Do I get that right and is that by design?

Reply:
yeah we store the buf in multiple places - but is just a pointer and a len which is very cheap to store (two u64 values) so its okay